Grammar.md

```diff
@@ -0,0 +1,53 @@
+## Grammar support
+
+Many languages - all Slavic (Russian, Ukrainian, Polish, Bulgarian, etc), Finnic (Finnish, Estonian) and others - have [grammatical case feature](https://en.wikipedia.org/wiki/Grammatical_case) that could be supported in OSRM Text Instructions too.
+Originally street names are being inserted into instructions as they're in OSM map - in [nominative case](https://en.wikipedia.org/wiki/Nominative_case).
+To be grammatically correct, street names should be changed according to target language rules and instruction context before insertion.
+
+Actually grammatical case applying is not the simple and obvious task due to real-life languages complexity.
+It even looks so hard so, for example, all known native Russian navigation systems don't speak street names in their pronounceable route instructions at all.
+
+But fortunately street names have restricted lexicon and naming rules and so this task could be relatively easily solved for this particular case.
+
+### Implementation details
+
+The quite universal and simplier solution is the changing street names with the prepared set of regular expressions grouped by required grammatical case.
+The required grammatical case should be specified right in instruction's substitution variables:
+
+- `{way_name}` and `{rotary_name}` variables in translated instructions should be appended with required grammar case name after colon: `{way_name:accusative}` for example
+- [languages/grammar](languages/grammar/) folder should contain language-specific JSON file with regular expressions for specified grammar case:
+```json
+{
+    "v5": {
+        "accusative": [
+            ["^ (\\S+)ая-(\\S+)ая [Уу]лица ", " $1ую-$2ую улицу "],
+            ["^ (\\S+)ая [Уу]лица ", " $1ую улицу "],
+            ...
+```
+- All such JSON files should be registered in common [languages.js](languages.js)
+- Instruction text formatter ([index.js](index.js) in this module) should:
+  - check `{way_name}` and `{rotary_name}` variables for optional grammar case after colon: `{way_name:accusative}`
```

Comment:
I like this syntax – it leaves open the possibility of additional transformations like capitalization, which could replace the "capitalizeFirstLetter" meta property currently set on some localizations.